xds/src/main/java/io/grpc/xds/OrcaUtil.java

```diff
@@ -0,0 +1,210 @@
+   * installed to receive callback when an per-request ORCA report is received.
+   *
+   * @param listener contains the callback to be invoked when an per-request ORCA report is
+   *     received.
+   */
+  public static ClientStreamTracer.Factory newOrcaClientStreamTracerFactory(
+      OrcaReportListener listener) {
+    return newOrcaClientStreamTracerFactory(NOOP_CLIENT_STREAM_TRACER_FACTORY, listener);
+  }
+
+  /**
+   * Creates a new {@link ClientStreamTracer.Factory} with provided {@link OrcaReportListener}
+   * installed to receive callback when an per-request ORCA report is received.
+   *
+   * @param delegate the delegate factory to produce other client stream tracing.
+   * @param listener contains the callback to be invoked when an per-request ORCA report is
+   *     received.
+   */
+  public static ClientStreamTracer.Factory newOrcaClientStreamTracerFactory(
+      ClientStreamTracer.Factory delegate, OrcaReportListener listener) {
+    return new OrcaClientStreamTracerFactory(delegate, listener);
+  }
+
+  /**
+   * Creates a new {@link LoadBalancer.Helper} with provided {@link OrcaReportListener} installed to
+   * receive callback when an out-of-band ORCA report is received.
+   *
+   * @param delegate the delegate helper that provides essentials for establishing subchannels to
+   *     backends.
+   * @param listener contains the callback to be invoked when an out-of-band ORCA report is
+   *     received.
+   * @param maxReportInterval the maximum expected interval of receiving periodical ORCA reports.
+   * @param unit time unit of {@code maxReportInterval} value.
+   */
+  public static LoadBalancer.Helper newOrcaReportingHelper(
+      LoadBalancer.Helper delegate,
+      OrcaReportListener listener,
+      long maxReportInterval,
+      TimeUnit unit) {
+    // TODO(chengyuanzhang): create a LoadReportingHelper that intercepts createSubChannel() method.
+    // This requires OOB ORCA service impl to be done first in order to create an OrcaClient.
+    return null;
+  }
+
+  /**
+   * The listener interface for receiving backend ORCA reports. The class that is interested in
+   * processing backend cost metrics implements this interface, and the object created with that
+   * class is registered with a component, using methods in {@link OrcaUtil}. When an ORCA report is
+   * received, that object's {@code onLoadReport} method is invoked.
+   */
+  public interface OrcaReportListener {
+
+    /**
+     * Invoked when an ORCA report is received.
+     */
+    void onLoadReport(OrcaLoadReport report);
+  }
+
+  @VisibleForTesting
+  static final class OrcaClientStreamTracerFactory extends ClientStreamTracer.Factory {
+
+    @VisibleForTesting
+    static final CallOptions.Key<OrcaReportBroker> ORCA_REPORT_BROKER_KEY =
+        CallOptions.Key.create("internal-orca-report-broker");
+    @VisibleForTesting
+    static final Metadata.Key<OrcaLoadReport> ORCA_ENDPOINT_LOAD_METRICS_KEY =
+        Metadata.Key.of(
```

Comment:
Wouldn't the server side also need this key somewhere?

Reply:
Depending on how our client and server packages will be structured. The server-side does have to have access to this key. It can duplicate the definition if it's not easy to share the key among client and server.